contrib/opencensus-ext-django/opencensus/ext/django/middleware.py

```diff
@@ -118,79 +109,28 @@ class OpencensusMiddleware(MiddlewareMixin):
-            self.sampler = self._sampler(_rate)
-        else:
-            self.sampler = self._sampler()
-
-        # Initialize the exporter
-        transport = convert_to_import(settings.params.get(TRANSPORT))
-
-        if self._exporter.__name__ == 'GoogleCloudExporter':
-            _project_id = settings.params.get(GCP_EXPORTER_PROJECT, None)
-            self.exporter = self._exporter(
-                project_id=_project_id,
-                transport=transport)
-        elif self._exporter.__name__ == 'ZipkinExporter':
-            _service_name = self._get_service_name(settings.params)
-            _zipkin_host_name = settings.params.get(
-                ZIPKIN_EXPORTER_HOST_NAME, 'localhost')
-            _zipkin_port = settings.params.get(
-                ZIPKIN_EXPORTER_PORT, 9411)
-            _zipkin_protocol = settings.params.get(
-                ZIPKIN_EXPORTER_PROTOCOL, 'http')
-            self.exporter = self._exporter(
-                service_name=_service_name,
-                host_name=_zipkin_host_name,
-                port=_zipkin_port,
-                protocol=_zipkin_protocol,
-                transport=transport)
-        elif self._exporter.__name__ == 'TraceExporter':
-            _service_name = self._get_service_name(settings.params)
-            _endpoint = settings.params.get(
-                OCAGENT_TRACE_EXPORTER_ENDPOINT, None)
-            self.exporter = self._exporter(
-                service_name=_service_name,
-                endpoint=_endpoint,
-                transport=transport)
-        elif self._exporter.__name__ == 'JaegerExporter':
-            _service_name = settings.params.get(
-                JAEGER_EXPORTER_SERVICE_NAME,
-                self._get_service_name(settings.params))
-            _jaeger_host_name = settings.params.get(
-                JAEGER_EXPORTER_HOST_NAME, None)
-            _jaeger_port = settings.params.get(
-                JAEGER_EXPORTER_PORT, None)
-            _jaeger_agent_host_name = settings.params.get(
-                JAEGER_EXPORTER_AGENT_HOST_NAME, 'localhost')
-            _jaeger_agent_port = settings.params.get(
-                JAEGER_EXPORTER_AGENT_PORT, 6831)
-            self.exporter = self._exporter(
-                service_name=_service_name,
-                host_name=_jaeger_host_name,
-                port=_jaeger_port,
-                agent_host_name=_jaeger_agent_host_name,
-                agent_port=_jaeger_agent_port,
-                transport=transport)
-        else:
-            self.exporter = self._exporter(transport=transport)
-
-        self.blacklist_hostnames = settings.params.get(
-            BLACKLIST_HOSTNAMES, None)
-
-        # Initialize the propagator
-        self.propagator = self._propagator()
+        settings = getattr(django.conf.settings, 'OPENCENSUS', {})
+        settings = settings.get('TRACE', {})
+
+        self.sampler = settings.get('SAMPLER', None) or \
```

Comment:
I missed it in #621, but are you doing `get(..., None) or foo` instead of `get(..., foo)` because `settings['SAMPLER']` might be an empty string?

Also do you expect settings values to be a mix of objects and config strings? This pattern might also get confusing if we ever store plain old strings in config and read them alongside these config strings that represent objects.

Reply:
I use `or` to prevent evaluation:

Constructor is called only when setting evaluates to False:
```
settings.get('SAMPLER', None) or always_on.AlwaysOnSampler()
```
Constructor is always called, which could be costly (depending on the actual implementation):
```
settings.get('SAMPLER', always_on.AlwaysOnSampler())
```

Reply:
Yeah, I was worried about this as well.
My current thinking is: for each configuration entry, we will have to design whether it is a string to be evaluated as a Python expression, or just a plain value. And this is why we have several `configuration.eval` calls (to make it explicit) instead of doing a `loop and eval`.

Reply:
What about this?

```python
try:
    sampler = configuration.load(settings['SAMPLER'])
except KeyError:
    sampler = always_on.AlwaysOnSampler()
```

That way the type check is gone and it's clear from the code that the setting is a string.

In any case you can replace `get(foo, None)` with `get(foo)` if you do keep the `get`s.

Reply:
Do we want the `SAMPLER` config to be a string, or it can be either a string or Exporter object?
I was following the original approach where the config could be either a string (where we need to construct the object based on the string) or a strong type.

The implementation in this PR is trying to support both:
```
'SAMPLER': 'opencensus.trace.samplers.always_on.AlwaysOnSampler()'
```
and
```
'SAMPLER': always_on.AlwaysOnSampler()
```

This gives us the benefit that if people use configuration in python source code, then can use the latter approach to benefit from static check (which seems to be common for Pyramid). If people loads the configuration from a JSON object or something similar, they can get dynamic evaluation (just install the ext, change config, no import required).
This comes with a cost of making the config/implementation a bit more complex. I'm open to take either approach.

Reply:
@c24t please let me know your thoughts.

I'll merge this PR now and work on the documentation / notes on breaking changes. I'm open to adjust the current implementation to make it better. (And I'm thinking if it is a good idea to extract the configuration logic into core opencensus lib, given there are a lot of duplicated logic in Flask/Django/Pyramid)